tox.ini

```diff
@@ -1,26 +1,121 @@
+    coverage_report
+    packaging
+
+skip_missing_interpreters = {tty:True:False}
+
+
+##
+# Build (default environment)
+##
 
 [testenv]
-changedir = .tox
-deps = -rrequirements-test.txt
-commands = coverage run --parallel --rcfile {toxinidir}/.tox-coveragerc -m pytest --doctest-modules {envsitepackagesdir}/hyperlink {posargs}
 
-# Uses default basepython otherwise reporting doesn't work on Travis where
-# Python 3.6 is only available in 3.6 jobs.
-[testenv:coverage-report]
-changedir = .tox
-deps = coverage
-commands = coverage combine --rcfile {toxinidir}/.tox-coveragerc
-           coverage report --rcfile {toxinidir}/.tox-coveragerc
-           coverage html --rcfile {toxinidir}/.tox-coveragerc -d {toxinidir}/htmlcov
+description = run tests
+
+basepython =
+    py26: python2.6
+    py27: python2.7
+    py34: python3.4
+    py35: python3.5
+    py36: python3.6
+    py37: python3.7
+    py38: python3.8
+    pypy: pypy
+    pypy3: pypy3
+
+deps =
+    test: coverage==4.5.4
+    test: idna==2.8
+    test: {py26,py27,py34}: pytest==4.6.6
+    test: {py35,py36,py37,py38}: pytest==5.2.1
+    test: pytest-cov==2.8.1
+
+passenv =
+    # See https://github.com/codecov/codecov-python/blob/master/README.md#using-tox
+    # And CI-specific docs:
+    #   https://help.github.com/en/articles/virtual-environments-for-github-actions#default-environment-variables
+    #   https://docs.travis-ci.com/user/environment-variables#default-environment-variables
+    #   https://www.appveyor.com/docs/environment-variables/
+    codecov: TOXENV CODECOV_* CI
+    codecov: GITHUB_*
+    codecov: TRAVIS TRAVIS_*
+    codecov: APPVEYOR APPVEYOR_*
+
+    # Used in our AppVeyor config
+    codecov: OS
+
+setenv =
+    test: COVERAGE_FILE={toxworkdir}/coverage.{envname}
+    {coverage_report,codecov}: COVERAGE_FILE={toxworkdir}/coverage
+    codecov: COVERAGE_XML={envlogdir}/coverage_report.xml
+
+commands =
+    test: pytest --cov=hyperlink --cov-report=term-missing:skip-covered --doctest-modules {posargs:hyperlink}
+
+
+##
+# Generate HTML coverage report
+##
+
+[testenv:coverage_report]
+
+description = generate coverage report
+
+basepython = python
+
+skip_install = True
 
+deps =
+    coverage==4.5.4
+
+commands =
+    coverage combine
+    coverage report
+    coverage html
+
+
+##
+# Publish to Codecov
+##
+
+[testenv:codecov]
+
+description = upload coverage to Codecov
+
+basepython = python
+
+skip_install = True
+
+deps =
+    coverage==4.5.4
+    codecov==2.0.15
+
+commands =
+    coverage combine
+    coverage xml -o "{env:COVERAGE_XML}"
+    codecov --file="{env:COVERAGE_XML}" --env \
+        GITHUB_REF GITHUB_COMMIT GITHUB_USER GITHUB_WORKFLOW \
+        TRAVIS_BRANCH TRAVIS_BUILD_WEB_URL TRAVIS_COMMIT TRAVIS_COMMIT_MESSAGE \
+        APPVEYOR_REPO_BRANCH APPVEYOR_REPO_COMMIT \
+        APPVEYOR_REPO_COMMIT_AUTHOR_EMAIL APPVEYOR_REPO_COMMIT_MESSAGE_EXTENDED
+
+
+##
+# Packaging
+##
 
 [testenv:packaging]
-changedir = {toxinidir}
```

Comment:
slightly concerned about dropping `changedir` here, because this prevents python from loading things in from the current path, as described here: https://blog.ganssle.io/articles/2019/08/test-as-installed.html

Reply:
For the packaging test, we aren't running the python code; we're running `setup.py`.  And `toxinidir` is already the default working directory.

For the tests… I was planning on following this up with the addition of a `src/` directory, which I think is sufficient for the stated purpose (and works with or without `tox`) and a known best practice.  I can put the `changedir` back, but I don't know what that would really add, and despite making the tax config a lot bigger already, I do prefer not to add things that aren't needed.

That said, I'm not all that invested in it if you think it's better to put it back.

Reply:
On further reflection, if I remove this, it should happen *after* the `src/` thing, so I'll stick it back in.

Reply:
Right… that also requires other path hooey elsewhere.  How about I just do the src/ thing?

Reply:
Haha src thing is fine by me, I think it's gotten popular enough now that most people can figure it out. And when in doubt there's links to send folks.